lib/posix/pthread.c

```diff
@@ -436,26 +428,37 @@ int pthread_join(pthread_t thread, void **status)
 		return ESRCH;
 	}
 
-	pthread_mutex_lock(&pthread->state_lock);
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
-	}
-
-	if (pthread->state == PTHREAD_EXITED) {
+	state_key = k_spin_lock(&pthread->state_lock);
+	switch (pthread->state) {
+	case PTHREAD_DETACHED:
+		ret = EINVAL;
+		break;
+	case PTHREAD_JOINABLE:
+		k_spin_unlock(&pthread->state_lock, state_key);
+		ret = k_thread_join(&pthread->thread, K_FOREVER);
+		__ASSERT(ret == 0, "k_thread_join() failed: %d", ret);
+		state_key = k_spin_lock(&pthread->state_lock);
+		if (ret != 0) {
```

Comment:
You might need to go back and re-validate the current `pthread->state` value to make sure the value is as expected; trying to reason about all possible transitions that can happen between the unlock and lock is hard.

Reply:
Nothing else would set the state here. It would actually be the responsibility of this function to set the state to `PTHREAD_EXITED`. However, since the state would be changed again only a few lines below, changing it here would be redundant.

I could add that line of code for clarity. Perhaps the compiler would optimize it out in some cases.

Reply:
You're unlocking the state_lock, so couldn't another thread call a thread API and mangle things before you manage to re-lock it?